File: src/resilience/preflight.ts
import { spawnCli } from '../workers/spawn-utils.js';
import { appendOperatorDegradedRecord } from '../orchestrator/ledger.js';
import { recordProxyProbe } from './proxy-state.js';

/**
 * Operator-role provider preflight (PRD §17 Phase 4, Q40).
 *
 * Operator loops (prd-curator.sh, po-loop.sh, selfbuild-loop.sh,
 * warroom-loop.sh, reviewer-loop.sh) must not dispatch an agent cycle while
 * the provider is hard-down: the loop either dies noisily mid-cycle or — the
 * 2026-09-02 curator failure class — completes "successfully" with a silent
 * noop. This gate reuses the orchestrate-loop probe pattern
 * (scripts/orchestrate-loop.sh): up to three cheap `omp -p OK` probes against
 * the configured worker CLI; on failure it records a structured
 * `operator-degraded` ledger row, updates the shared circuit state
 * (.devagent/proxy-state.json via recordProxyProbe), and reports a decision
 * the caller must honor by skipping that cycle's agent dispatch.
 */

/** Roles a preflight can gate; each maps to one operator loop script. */
export const PREFLIGHT_ROLES = ['prd-curator', 'po', 'selfbuild', 'warroom', 'reviewer'] as const;

export type PreflightRole = (typeof PREFLIGHT_ROLES)[number];

export function isPreflightRole(value: string): value is PreflightRole {
  return (PREFLIGHT_ROLES as readonly string[]).includes(value);
}

/** Probe invocations per gate run. */
export const PREFLIGHT_PROBE_ATTEMPTS = 3;

/** Hard wall-clock cap per probe. Must stay cheap but clear the slowest
 * observed gateway round-trip: omniroute/dev replies land at 25-38s (2026-09-03
 * live: shell probe 26s, execFile 25s, several >30s), so a 30s cap turned the
 * gate into a coin-flip that tripped the selfbuild circuit breaker. 60s keeps
 * the probe bounded while clearing the tail. */
export const PREFLIGHT_PROBE_TIMEOUT_MS = 60_000;

/** Sleep between failed probes (mirrors orchestrate-loop's 5s). */
export const PREFLIGHT_RETRY_DELAY_MS = 5_000;

/** The exact prompt every probe sends. */
export const PREFLIGHT_PROBE_PROMPT = 'OK';

/** Stable ledger taskId for operator preflight rows. */
export const PREFLIGHT_LEDGER_TASK_ID = 'operator-preflight';

/** One probe outcome. */
export interface PreflightProbe {
  ok: boolean;
  /** Bounded excerpt of the failure (stderr/stdout tail) when not ok. */
  detail?: string;
}

/** Final gate decision. */
export interface PreflightDecision {
  /** true = proceed with the cycle's agent dispatch. */
  ok: boolean;
  role: PreflightRole;
  attempts: number;
  /** Worker CLI probed, when the caller declared it (repo config). */
  worker?: string;
  /** Model id probed, when the caller declared it (empty = CLI default). */
  model?: string;
  /** Bounded last-failure excerpt; present when the gate degraded. */
  detail?: string;
}

/** Bound an error excerpt for ledger/log output. */
function boundDetail(text: string): string {
  return text.replace(/\s+/g, ' ').trim().slice(0, 200);
}

/**
 * One provider probe: ask the worker CLI to reply to "OK" and require an
 * answer. `--mode json` success looks like an event stream containing
 * `"text":"OK"`; a probe that exits 0 without it is degraded. This mirrors
 * the orchestrate-loop probe but runs through the same spawn path as the
 * worker adapters so env hardening (nested-env blocklist, PATH fallback)
 * stays consistent.
 */
export async function runPreflightProbe(
  cmd: string,
  args: string[],
  opts: { cwd: string; env?: Record<string, string>; /** Wall-clock cap; default PREFLIGHT_PROBE_TIMEOUT_MS. */ timeoutMs?: number },
): Promise<PreflightProbe> {
  const run = await spawnCli(cmd, args, { cwd: opts.cwd, timeoutMs: opts.timeoutMs ?? PREFLIGHT_PROBE_TIMEOUT_MS, env: opts.env });
  const ok = run.exitCode === 0 && run.stdout.includes('"text":"OK"');
  return ok
    ? { ok: true }
    : { ok: false, detail: boundDetail(`${run.stderr}\n${run.stdout}`) || `exit ${run.exitCode}` };
}

/**
 * The typed gate. Runs up to PREFLIGHT_PROBE_ATTEMPTS probes with
 * PREFLIGHT_RETRY_DELAY_MS between failures, records the circuit transition
 * (recordProxyProbe) and — on failure — one structured `operator-degraded`
 * ledger row, then decides. Callers MUST skip the cycle's agent dispatch
 * when `decision.ok` is false; that skip is the visible degradation.
 */
export async function runPreflightGate(args: {
  repoPath: string;
  role: PreflightRole;
  /**
   * Probe argv WITHOUT the prompt: `["<cmd>", "-p", ...flags]`. The gate owns
   * prompt placement — the prompt immediately follows the `-p` flag, then
   * caller flags, matching buildOmpArgs and the orchestrate-loop probe (a
   * trailing prompt would be swallowed as `-p`'s value).
   */
  argv: string[];
  cwd?: string;
  /** Worker CLI being probed (repo config; recorded on the ledger row). */
  worker?: string;
  /** Model id being probed (repo config; empty string = CLI default). */
  model?: string;
  /** Injection seam for tests. */
  probe?: (cmd: string, a: string[], o: { cwd: string; env?: Record<string, string> }) => Promise<PreflightProbe>;
  /** Injection seam for tests: sleep between failed probes. */
  delayMs?: (ms: number) => Promise<void>;
}): Promise<PreflightDecision> {
  const probe = args.probe ?? runPreflightProbe;
  const delayMs = args.delayMs ?? ((ms: number) => new Promise<void>((r) => setTimeout(r, ms)));
  const cwd = args.cwd ?? args.repoPath;
  const [cmd, promptFlag, ...flags] = args.argv;
  if (!cmd || promptFlag !== '-p') {
    throw new Error('preflight argv must start with ["<cmd>", "-p", ...flags] (prompt flag before flags)');
  }

  let attempts = 0;
  let last: PreflightProbe = { ok: false, detail: 'no probe ran' };
  while (attempts < PREFLIGHT_PROBE_ATTEMPTS) {
    attempts += 1;
    try {
      last = await probe(cmd, [promptFlag, PREFLIGHT_PROBE_PROMPT, ...flags], { cwd });
    } catch (err) {
      last = { ok: false, detail: boundDetail((err as Error).message) };
    }
    if (last.ok) break;
    if (attempts < PREFLIGHT_PROBE_ATTEMPTS) await delayMs(PREFLIGHT_RETRY_DELAY_MS);
  }

  const decision: PreflightDecision = {
    ok: last.ok,
    role: args.role,
    attempts,
    ...(args.worker ? { worker: args.worker } : {}),
    ...(args.model !== undefined ? { model: args.model } : {}),
    ...(last.ok ? {} : { detail: last.detail }),
  };
  if (last.ok) {
    // Advance the shared circuit on success too: without this a provider
    // recovery leaves a stale `open` from the last failure, and consumers
    // reading the circuit keep short-circuiting (2026-09-03: circuit stuck
    // open since 13:21Z despite green probes from 01:58Z).
    recordProxyProbe(args.repoPath, { ok: true });
  } else {
    // Shared circuit state so `devagent status --providers` reports the
    // operator-loop degradation exactly like the orchestrate-loop gate.
    recordProxyProbe(args.repoPath, { ok: false, detail: `preflight[${args.role}]: ${last.detail ?? 'failed'}` });
    // Structured degradation row: the ledger is the evidence a degraded
    // factory cycle was skipped, not silently noop'd (Q40).
    appendOperatorDegradedRecord(args.repoPath, {
      ts: new Date().toISOString(),
      kind: 'event',
      event: 'operator-degraded',
      taskId: PREFLIGHT_LEDGER_TASK_ID,
      attempt: attempts,
      role: args.role,
      ok: false,
      attempts,
      worker: args.worker ?? '',
      model: args.model ?? '',
      ...(last.detail ? { detail: last.detail } : {}),
    });
  }
  return decision;
}

File: internal/resilience/preflight.go
// Package resilience seeds the Go port of src/resilience/preflight.ts
// (Q40). Only the single probe function lands here with FR-GO-02 — the
// typed gate, ledger rows, circuit, and pager arrive with the orchestrator
// port (FR-GO-07, issue #194).
package resilience

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/FreePeak/devagent/internal/spawn"
)

// One probe outcome.
type Probe struct {
	OK     bool
	Detail string
}

// boundDetail bounds an error excerpt for ledger/log output.
func boundDetail(text string) string {
	return trunc(strings.Join(strings.Fields(text), " "), 200)
}

func trunc(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

// OMPStartupWedgePattern: omp's own startup-watchdog signature
// ("Still starting after <n>s — phase: ...").
var OMPStartupWedgePattern = regexp.MustCompile(`Still starting after \d+s`)

// probeMarkerSeen reports whether the probe answer has already streamed:
// `--mode json` (omp) success looks like an event stream containing
// `"text":"OK"`; grok's `--output-format streaming-json` emits the answer as
// a text chunk — `{"type":"text","data":"OK"}` (captured 2026-09-06) — so
// both shapes count.
func probeMarkerSeen(stdout string) bool {
	return strings.Contains(stdout, `"text":"OK"`) ||
		strings.Contains(stdout, `"type":"text","data":"OK"`)
}

// RunPreflightProbe asks the worker CLI to reply to "OK" and require an
// answer. It completes on the STREAMED MARKER, not the process exit
// (issue #308): measured 2026-09-11 on the live route, the answer arrived at
// ~24s (`stopReason:"stop"`, ttft 8.1s) with `"text":"OK"` present 3x in the
// captured stdout, while the CLI process still lingered past a 70s kill —
// waiting on exit turned a HEALTHY provider into provider-degraded rows, the
// breaker, and endless hub restarts. RunCliUntil kills the process tree the
// moment the predicate fires; the predicate, not the exit code, is the
// verdict. A run that ends without ever streaming the marker (exit 0,
// nonzero exit, or deadline) is degraded. This mirrors the orchestrate-loop
// probe but runs through the same spawn path as the worker adapters so env
// hardening stays consistent.
func RunPreflightProbe(cmd string, args []string, dir string, timeoutMs int) Probe {
	if timeoutMs <= 0 {
		timeoutMs = 60_000
	}
	run := spawn.RunCliUntil(cmd, args, spawn.Options{Dir: dir, TimeoutMs: timeoutMs}, probeMarkerSeen)
	if probeMarkerSeen(run.Stdout) {
		return Probe{OK: true}
	}
	detail := boundDetail(run.Stderr + "\n" + run.Stdout)
	if detail == "" {
		detail = "exit " + strconv.Itoa(run.ExitCode)
	}
	return Probe{OK: false, Detail: detail}
}
